### ISSUE TITLE
Suppress UnusedFunctionParameter with underscore prefix

### ISSUE THREAD

**Issue by glen-84** [squizlabs/PHP_CodeSniffer#1962]
In cases where you need some of the parameters, but not all of them, for example:

```php
// @codingStandardsIgnoreLine -- Not using the $slug parameter.
public function stageAction($id, $slug, $stageId)
```

TypeScript allows you to use an underscore prefix to suppress the error/warning, like this:

```php
public stageAction(id, _slug, stageId)
```

Or, in PHP:

```php
public function stageAction($id, $_slug, $stageId)
```

It's a lot cleaner and less verbose in my opinion.

The variable name sniff would need an option to allow a single underscore in front of function parameters.

**Comment by gsherwood** [comment]
I feel like this would be far too specific to put into a core sniff because it's not an agreed standard in the PHP community. Ill leave this open to see if it gets any more support and I'm just not aware of how wide spread it is.

**Comment by jrfnl** [comment]
PR #1318 proposes changes to this sniff which would allow @glen-84 to suppress the error by excluding/including certain an error codes.

I do think the PR needs some more work, but need some guidance on how to polish it off (was one of my earlier PRs to this repo).

**Comment by jrfnl** [comment]
@glen-84 PR #1318 has been merged and released as part of PHPCS 3.4.0. Did that sufficiently address this issue by allowing you to exclude certain error codes ?

**Comment by mwgamble** [comment]
It looks like with the updates to this sniff, it assumes that if there is an unused parameter in a method in a class that implements an interface, the reason for the parameter being unused must be because the method is part of the interface being implemented.

However, this isn't necessarily the case. I have multiple instances of classes that implement an interface, but also have additional methods that aren't part of the interfaces. These methods are all supplied with a second parameter that is unused, but can be used by subclasses that override the relevant methods.

While I could technically suppress the `FoundInImplementedInterfaceAfterLastUsed` code for these methods, it feels wrong, as that's not actually the issue. I feel like it would be a lot simpler if I could just suppress the warnings by following the convention set out in other languages - prefixing the variable name with an underscore.

**Comment by glen-84** [comment]
> @glen-84 PR #1318 has been merged and released as part of PHPCS 3.4.0. Did that sufficiently address this issue by allowing you to exclude certain error codes ?

Unfortunately not quite. The [question](https://gitter.im/squizlabs/PHP_CodeSniffer?at=5c309b13357dd7655d1ea591) that I asked on Gitter was related to what @mwgamble describes above (there are probably a few variations – class vs interface, etc.).

**Comment by DMGPage** [comment]
I would like to see any solution of this problem. I encounter it too often in our project. And it's not about interface. Even if I overwrite parent class function, I have to make it compatible with parent function params. Even if I do not use them. And second problem is anonymous function. Most of the time I can't change priority how params is passed. And if I need only last param, I have to write all of them. So, everywhere I have no other option, but write ignore code "phpcs:ignore". Sometimes it's ok, but for overwriting method it's not working. If I write this comment after method description, IDE won't see description. I could write "phpcs:disable" and "phpcs:enable", but this will lead to more mess in code. Need some better solution for this.

**Comment by serious-angel** [comment]
Wonderful day! Is there any mind change on it? My gracious sake, does it feel up the warning/problem list in the IDE when it's not required!

It's quite frequent when you pass a callable as an argument that receives at least two arguments, and you need the second only.

For example, in Laravel ServiceProviders:

```php
class ProtectionServiceProvider extends ServiceProvider implements DeferrableProvider
{
    public function register(): void
    {
        // phpcs: SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter: Unused parameter $app.
        $this->app->bind(Some::class, function ($app, array $parameters = []) {
            return new Some(
                prefix: $parameters['prefix'] ?? 'default'
            );
        });

        // phpcs: SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter: Unused parameter $_app.
        $this->app->bind(Some2::class, function ($_app, array $parameters = []) {
            return new Some2(
                cache: cache('some_store'),
                prefix: $parameters['prefix'] ?? null
            );
        });
    }
}
```

Yet, there are projects with around 100 different bindings/singletons, where just a few utilize the `$app` property, or use `app()` instead at all.

---

Funny, or not... but some LLMs (e.g. Gemini Flash 3) outputted the following nonsense:

<details>
<summary>"ignoredAndReferencedNames"</summary>
To fix your specific error, ensure you are using the full sniff name in the ref attribute and placing the property correctly:  

```xml
<rule ref="SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter">
    <properties>
        <property name="ignoredAndReferencedNames" type="array">
            <element value="$app"/>
        </property>
    </properties>
</rule>
```
</details>

<details>
<summary>"ignoreUnderscorePrefix"</summary>
Add this block to your phpcs.xml file. Note the specific sub-sniff string:  
  
```xml
<rule ref="SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter">
    <properties>
        <property name="ignoreUnderscorePrefix" type="boolean" value="true"/>
    </properties>
</rule>
```
Why it likely failed before
Boolean vs String: Many people try to use value="1" or omit the type="boolean". Slevomat is often strict about the XML type definition.

The Full Path: If you only use ref="SlevomatCodingStandard.Functions.UnusedParameter", the property might not "trickle down" to the actual check logic. Using the triple-name (...UnusedParameter.UnusedParameter) targets the logic directly.
</details>
